Core/clim-basic/geometry/region-predicates.lisp

```diff
@@ -66,12 +77,52 @@
   (every (lambda (r) (region-contains-position-p r x y))
          (standard-region-set-regions self)))
 
-(defmethod region-contains-position-p ((self standard-region-difference) x y)
-  (and (region-contains-position-p (standard-region-difference-a self) x y)
-       (not (region-contains-position-p (standard-region-difference-b self) x y))))
+(defmethod region-contains-position-p ((region standard-region-difference) x y)
+  (let ((region-a (standard-region-difference-a region))
+        (region-b (standard-region-difference-b region)))
+    (and (region-contains-position-p region-a x y)
+         (not (region-contains-position-p region-b x y)))))
+
+(defmethod region-contains-position-p ((self point) x y)
+  (and (coordinate= (point-x self) x)
+       (coordinate= (point-y self) y)))
 
-(defmethod region-contains-position-p ((self standard-polygon) x y)
-  (and (region-contains-position-p (bounding-rectangle self) x y)
+(defmethod region-contains-position-p ((self polyline) x y)
+  (setf x (coordinate x)
+        y (coordinate y))
+  (block nil
+    (map-over-polygon-segments
+     (lambda (x1 y1 x2 y2)
+       (when (segment-contains-point-p x1 y1 x2 y2 x y)
+         (return t)))
+     self)
+    nil))
+
+(defmethod region-contains-position-p ((self line) x y)
+  (multiple-value-bind (x1 y1) (line-start-point* self)
+    (multiple-value-bind (x2 y2) (line-end-point* self)
+      (segment-contains-point-p x1 y1 x2 y2 x y))))
+
+(defmethod region-contains-position-p ((self elliptical-arc) x y)
+  (nest
+   (multiple-value-bind (cx cy) (ellipse-center-point* self))
+   (let ((alpha (ellipse-start-angle self))
+         (omega (ellipse-end-angle self))
+         (polar->screen (polar->screen self)))
+     (and (angle-contains-point-p alpha omega (- x cx) (- y cy))
+          (nest (multiple-value-bind (polar-x polar-y)
+                    (untransform-position polar->screen x y))
+                ;; FIXME we don't need to factor the additonal epsilon
+                ;; but rotated elliptoids are naively rendered in clx.
+                (multiple-value-bind (polar-dx polar-dy)
+                    (untransform-distance polar->screen 1 1))
+                (coordinate-between*
+                 (- 1 (+ (square polar-dx) (square polar-dy)))
+                 (+ (square polar-x) (square polar-y))
+                 (+ 1 (+ (square polar-dx) (square polar-dy)))))))))
+
+(defmethod region-contains-position-p ((region polygon) x y)
+  (and (region-contains-position-p (bounding-rectangle region) x y)
```

Comment:
Would it make sense to do this generally in an `:around` method for 2-dimensional regions like for `region-intersects-region-p`?

Reply:
Maybe specialized on `area` or the cached-bounding-rectangle thing.

Reply:
it does make sense but not as things are now. I have plans to cache bounding rectangle for all regions but as things are now it would be very inefficient i.e for ellipses.

region-intersects-region-p around method is there to provide a negative test, because next method (region-intersection) is not defined for all pairs of geometrical objects. In other words it is a hack which makes us fail less often with no-applicable-method exception or with incorrect result from blind creation of standard-region-intersection object when we do not know what to do.

Reply:
OK, I guess we can improve this later.